tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -12,889 +13,945 @@
+            // If the type defines multiple OnMethodEnd methods to work with both void return types and non-void return types,
+            // iterate over the methods to disambiguate
+            var possibleMethods = integrationType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+            for (int i = 0; i < possibleMethods.Length; i++)
             {
-                // If the type defines multiple OnMethodEnd methods to work with both void return types and non-void return types,
-                // iterate over the methods to disambiguate
-                var possibleMethods = integrationType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
-                for (int i = 0; i < possibleMethods.Length; i++)
-                {
-                    var possibleMethod = possibleMethods[i];
+                var possibleMethod = possibleMethods[i];
 
-                    if (possibleMethod.Name == EndMethodName && possibleMethod.ReturnType.Name == returnTypeName)
-                    {
-                        return possibleMethod;
-                    }
+                if (possibleMethod.Name == EndMethodName && possibleMethod.ReturnType.Name == returnTypeName)
+                {
+                    return possibleMethod;
                 }
             }
-
-            return null;
         }
 
-        private static void WriteCreateNewProxyInstance(ILGenerator ilWriter, Type proxyType, Type targetType)
-        {
-            ConstructorInfo proxyTypeCtor = proxyType.GetConstructors()[0];
+        return null;
+    }
 
-            if (targetType.IsValueType && !proxyTypeCtor.GetParameters()[0].ParameterType.IsValueType)
-            {
-                ilWriter.Emit(OpCodes.Box, targetType);
-            }
+    private static void WriteCreateNewProxyInstance(ILGenerator ilWriter, Type proxyType, Type targetType)
+    {
+        var proxyTypeCtor = proxyType.GetConstructors()[0];
 
-            ilWriter.Emit(OpCodes.Newobj, proxyTypeCtor);
+        // No null check is needed for value types
+        if (targetType.IsValueType && !proxyTypeCtor.GetParameters()[0].ParameterType.IsValueType)
+        {
+            ilWriter.Emit(OpCodes.Box, targetType);
         }
 
-        private static TTo UnwrapReturnValue<TFrom, TTo>(TFrom returnValue)
-            where TFrom : IDuckType
+        ilWriter.Emit(OpCodes.Newobj, proxyTypeCtor);
+    }
+
+    private static TTo? UnwrapReturnValue<TFrom, TTo>(TFrom? returnValue)
+        where TFrom : IDuckType
+    {
+        if (returnValue is not null)
         {
             return returnValue.GetInternalDuckTypedInstance<TTo>();
         }
 
-        private static async Task<TTo> UnwrapTaskReturnValue<TFrom, TTo>(Task<TFrom> returnValue, bool preserveContext)
-            where TFrom : IDuckType
+        Log.Debug("UnwrapReturnValue<{TFrom}, {TTo}>: The return value is null.", typeof(TFrom), typeof(TTo));
+        return default;
```

Comment:
This is the real "bug fix" part of this PR, right? Is it worth adding a test for this behaviour, to confirm it's definitely fixed (and we don't regress?)